### ISSUE TITLE
Support two-stage parameter grid search

### ISSUE THREAD

**Issue by agitter** [Reed-CompBio/spras#318]
A common pathway reconstruction workflow is to start with an initial grid of parameter combinations, inspect the output pathways by applying heuristics, and filter the output pathways that do not fall within the desirable heuristic ranges. An optional final step could be to refine the initial grid of parameter combinations to remove those that produced undesirable outputs and provide more granular step sizes around parameter values that produced desirable outputs.

SPRAS could directly support this with a few new features:
- Expanding the graph properties (heuristics) computed in the summary table. This could include graph statistics like clustering coefficient, etc.
- Allow a user to specify ranges of graph properties (heuristics) they find tolerable in the config file
- Filter output pathways based on tolerable ranges
- Rewrite a new config file with the parameter combinations that produced those tolerable ranges so the user can use it as a template for a second run

An example of filtering from https://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1011492:
> After generating subnetworks for each parameter combination, we discarded: any empty networks; networks that did not have a connected component with at least 25 vertices; networks in which one vertex was connected to 10% or more of the total vertices (which indicates a dominant hub node); and networks that contained a chain of more than 8 vertices without any nonzero prizes. An additional filter for networks whose vertices were not at least half prize nodes did not affect any of the networks.

**Comment by ntalluri** [comment]
<img width="969" height="587" alt="Image" src="https://github.com/user-attachments/assets/482bd5f6-5401-482e-b309-ce6d384127a3" />

**Comment by tristan-f-r** [comment]
Filtering will be manual, but I think in that comment I was mainly curious about just the first step: how will the grid be generated in the first place?

**Comment by ntalluri** [comment]
I will be predefining the first coarse grid, but I'm still deciding what that grid should be. I'm assuming I can figure it out when I do the performance testing of the parameters, otherwise I will set a grid per algorithm based on my knowledge of the parameters.

**Comment by ntalluri** [comment]
This document currently contains the new idea for two stage parameter tuning: 

https://docs.google.com/document/d/1PJFGeNkE63RkopGdRIkEJYOJ2yPFKkxeIMtqXANep6U/edit?usp=sharing

**Comment by ntalluri** [comment]
I have a couple of questions/comments:

1) for the metadata to be added to individual algorithm parameters, what is this supposed to be? Are these just algorithm parameters or other information now?

2) For the dataset category and per dataset heuristics, is the goal for these so we can put every dataset category/dataset in a single configuration file for stage 1 to run the coarse grid search? 

3) I'm confused why there is a general heuristics vs dataset level version of heuristics; I'm confused on the purpose of having both. Could you explain more on why we need both? (I can understand have one or the other though). 
- Since the dataset category could be implemented, another idea could be to set general hueristics that are set per dataset category

```
label: [Panther Pathways]
huerisitcs: [nodes < 500, edges < 1000, componets = 1 ...]
label: [GWAS]
huerisitcs: [nodes < 700, edges < 50000, componets = 3 ...]
... (for the other dataset categories)
```

4) Could you explain further how this idea will finetune per dataset within a dataset category? I don't see that idea written for stage 2 other than making the new ranges but those seem to be a single range that is applied per dataset category configuration.

**Comment by tristan-f-r** [comment]
1. With #329, we can take any parameter (say, pathlinker's `k`) and give it metadata [the range for the coarse grid run] right in its schema. 
2. Yes. Also for the stage 2 grid search, so we can have everything in one config file instead of splitting the config files into each dataset category. The motivation for this will be in (6).
3. This might be a misunderstanding on my part: I'll reread that and make sure that I understand the intent of where heuristics were to be placed (if so, I believe I accidentally misread a set of 'general' heuristics as well as a set of per-dataset ones: since the general heuristics don't exist, it would just be the per-dataset heuristics.) Your general config format there makes sense 👍 
4. We can! This is one reason why I was hoping for everything to be in the same config file for Stage 1, that way it's very easy to update heuristics and all of the cache information can be used instead of moving it around. (I assume that under different configs, we will have different output folders)
5. This implementation isn't written, though it's mainly because I'm not worried about its implementation. [I'm much more worried about the parts around parameter tuning]
6. Generally, I think its possible to do HTCondor run jobs and okay. It just seems that if we do multiple HTCondor long jobs, we lose good monitoring for monitor multiple long condor jobs and testability: it's much easier to test SPRAS features than HTCondor ones.
7. This will most likely be a truncated form of the SPRAS config, with just `dataset` and `analysis`, where the final config will be generated with the SPRAS CLI.

**Comment by ntalluri** [comment]
The way I was thinking about heuristics is that they will be defined per dataset collection. Each different collections can have different heuristics but they are set at the collection level not the dataset level.

In my head these would be set to what I wrote above:
```
label: [Panther Pathways]
heuristics: [nodes < 500, edges < 1000, componets = 1 ... (other graph topological huersitics)]
label: [GWAS]
heuristics: [nodes < 700, edges < 50000, componets = 3 ...]
... (for the other dataset categories)
```
And when we do stage 2 (fine tuning) these heuristics would stay exactly the same.

The confusion might have come from the idea of splitting into separate configs per dataset; what I meant was that the heuristics defined in the new config for each dataset would be the heuristics corresponding to dataset collection it came from (sorry for the confusion!).

**Comment by ntalluri** [comment]
Based on the conversation we had during our meeting, we should be aiming to do fine tuning per dataset in it's own configuration file.
- This is also dependent on having a well running htcondor plugin
- Also we might need to have a better way to communicate / monitor all of the long jobs we will be running.

In Stage 1, we’ll run a coarse grid search using a single configuration file that defines a collection of datasets.
In Stage 2, we’ll generate a fine-tuned grid for each individual dataset in their own config file, using the good parameter combinations from the coarse grid as starting points before expanding the search space.

**Comment by tristan-f-r** [comment]
I can put a more concrete plan now which can dramatically help, focusing on:

> this seems like something that we can figure out when we get there: especially with dependent parameter combinations, this seems like it has too many smaller edge cases to address in the overall plan. For now, we can assume that this bridge between the first and second grid search exists.

For generating a fine-tuned grid from a coarse-tuned grid, we can actually do the following:

When generating the config, given the coarse-tuned grid G, we can generate a secondary intermediary grid G' such that G' is G but with lower step values (for example, by cutting the step size in half, the 3d cube example mentioned in the recent meeting will have an intermediary grid containing the corners, parameters in the middle of each edge, and a parameter at the center of the cube).

Then, for each point in G', we give that parameter run a dependency: that run will only happen _if_ its closest neighbors in G were successful.

We embed these dependencies into the configuration, allowing us to run the first and second stage of parameter tuning _in parallel_.

For example, take the very simple case of a single algorithm with two parameters `a` and `b`:

If the coarse grid was something like

```
(0, 0)       (1, 0)


(0, 1)       (1, 1)
```

We let our intermediary grid be


```
(0, 0)   (0.5, 0)    (1, 0)

(0, 0.5) (0.5, 0.5)  (1, 0.5)

(0, 1)   (0.5, 1)    (1, 1)
```

where:
- `(0.5, 0)` depends on either `(0, 0)` or `(1, 0) (that parameter combination will run if one of those parameter combinations are successful)
- `(0, 0.5)` depends on `(0, 1)`, `(0, 0`)
- ...
- `(0.5, 0.5)` depends on `(0, 0)`, `(1, 0)`, `(0, 1)`, `(1, 1)`

If this algorithm had varying runtime (which is the case for algorithms like pathlinker), we can actually parallelize results, meaning it would be faster to have one configuration here rather than having several configurations.

**Comment by tristan-f-r** [comment]
## First grid search

We need a way to be able to:

1. Attach range and step information to parameters: each parameter should have a (coarse) range of parameters that works for datasets in general. #329 gives us the ability to attach metadata to individual algorithm parameters. (We will need a follow-up PR to enable range metadata)
2. Have a list of graph heuristics
    * Graph heuristics are specified in general for a configuration file. This will be its own PR that allows runs to fail (as clarified in (4)) when they don't meet their dataset's category's graph heuristics.
3. Generate a parameter config using the ranges from (1):
    * This will be done with the SPRAS CLI (#332), where we input our `datasets` (which also happen to contain our general graph heuristics from (2))
    * This will parse the internal ranges with #329, after making a PR building on top of it that adds range union support.
4. Get a list of parameter combinations that 'succeeded'
    * We say that a run fails if it doesn't meet the graph heuristic criteria. This allows us to dump all of the errors we get (dataset incompatibility errors, internal SPRAS errors, and our new graph heuristic errors) all in the same place, and make our grid search more tolerant of errors in general. We can also better classify them (see #394) to make it easier on the user to spot issues with SPRAS/their configuration/actual graph heuristic problems.

We can dump some file (as suggested by @ntalluri) in `logs` of all of the jobs and if they had errors or not (i.e. a `report.json` that can later be formatted as proposed in #380.).

## Second grid search

Using the information in the `report.json` file, we can generate a new configuration file (again with the SPRAS CLI), containing all of the parameters 'in between' of our successful runs by incrementing the steps of the ranges parsed in numeric values (enum values will not get more fine-grained than their original input).

We'll set certain parameter combinations to only run on certain datasets by generating many configuration files. (If we want cross-dataset analysis, this can still use #406 instead.)


From here, we move to evaluation of all of our parameter runs, where, under my impression, this part has already been mostly implemented.

---

**Comment by tristan-f-r** [comment]
This implementation plan can be extended for nondeterminism support for OSDF integration (we can have some kind of set seed for a seed factory in the initial configuration which generates the tuned configurations), but it might just be better to rewrite it at that point.

**Comment by tristan-f-r** [comment]
[The latest comment doesn't utilize a single configuration file.](https://github.com/Reed-CompBio/spras/issues/318#issuecomment-3411590522) As I briefly commented, we might need to change the workflow later to be in a single configuration file if we want this to work with OSDF, but @ntalluri pointed out that we don't need this for the SPRAS paper.

A union of ranges is good when we need to exclude certain ranges. (Regardless, we can't _only_ use ranges).

**Comment by tristan-f-r** [comment]
> What does it mean for a run to fail if the graph heuristics aren't met? I view unsatisfied graph heuristics differently from a failure to execute the algorithm successfully or failure to provide valid input. I assumed we should still output pathway files for parameter combinations that don't meet the graph heuristics.

@ntalluri was asking about this as well, and I think it might be beneficial to just draft a PR for it: the general idea is that we should have different error types (timeout errors, compatibility errors, 'graph heuristic fail' errors), but we preserve whatever output we get. The implementation ideas here are specific enough that it might just be better to write it as code, since it eventually just relies on what we're able to do with Snakemake.
